Initial version of spezi-firebase-template

PaulGoldschmidt opened the pull request:
# Initial version of spezi-firebase-template
This PR merges the effort of developing a unified spezi firebase template repository into a first working state.
See the descriptions of the individual PRs that were merged into this repo as a more in-depth description of what the functions are. The next steps are documented in the issues. Idea behind merging this now is that we can check the full CI/CD lifecycle. Full documentation and function finetuning will be provided with the next PRs. Multiple approaches choosen (testing via vitest, automatic package version sync & emulator setup) are novel in the SPEZI Ecosystem and serve as a blueprint for future efforts in the ecosystem itself; further, the template uses ecosystem provided packages (linting, FHIR models) as foundations to lead the effort in unifying often used system components.

Thanks for the support here, @arkadiuszbachorski @pauljohanneskraft & @PSchmiedmayer 🚀

## :pencil: Code of Conduct & Contributing Guidelines
By creating and submitting this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Cloud functions for step counts, user data retrieval, device register/unregister, messaging (add/dismiss), welcome messages, and user administration.

* **Infrastructure & Security**
  * Firebase project and emulator setup, Firestore and Storage security rules, and CI/CD workflows for build, tests, scanning, and deployment (including scheduled checks).

* **Documentation**
  * Expanded README, license and reuse metadata, and formatting/linting configs.

* **Tests**
  * Comprehensive unit/integration tests and test helpers for functions and auth.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

PaulGoldschmidt committed:
init

PaulGoldschmidt committed:
introduction of spezi-web-configurations

PaulGoldschmidt committed:
linter issues fixed

PaulGoldschmidt committed:
Update package.json

PaulGoldschmidt committed:
emulator setup

PaulGoldschmidt committed:
emulator setup

PaulGoldschmidt committed:
adopt @stanfordspezi/spezi-web-configurations

PaulGoldschmidt committed:
Update package.json

PaulGoldschmidt committed:
add spezi-firebase-utils, spezi-firebase-fhir & spezi-firebase-cloud-messaging

PaulGoldschmidt committed:
migrate to spezi-firebase-fhir fully

PaulGoldschmidt committed:
update functions to use spezi-firebase-fhir

PaulGoldschmidt commented:
Quick overview over the current template status:

1. Basic setup & npm configuration done
2. Basic functionality of functions (especially `onUserCreated`) implemented
3. out-of-the-box compatibility with [StanfordSpezi/spezi-web-configurations](https://github.com/StanfordSpezi/spezi-web-configurations/) planed and basic integration done, but [Issue #14](https://github.com/StanfordSpezi/spezi-web-configurations/issues/14) will need to be closed beforehand to be usable here
4. clinican/patient/admin-model adopted
5. Our packages from [StanfordSpezi/spezi-firebase](https://github.com/StanfordSpezi/spezi-firebase) (@stanfordspezi/spezi-firebase-utils, @stanfordspezi/spezi-firebase-cloud-messaging & @stanfordspezi/spezi-firebase) are integrated by default

Next steps:

- [x] Adress [Issue #14](https://github.com/StanfordSpezi/spezi-web-configurations/issues/14) in [StanfordSpezi/spezi-web-configurations](https://github.com/StanfordSpezi/spezi-web-configurations/)
- [x] Add workflows for CI/CD
- [x] Testing (😱)
- [x] more basic functions
- [x] `npm run serve:seeded` should actually seed data (shocker, ikik)

I would suggest discussing the full scope of the templete next week, what do you think @pauljohanneskraft @arkadiuszbachorski? (cc: @PSchmiedmayer)

PSchmiedmayer commented:
Thank you for working on this @PaulGoldschmidt. Happy to setup a short meeting, maybe Friday this week. Let's sync on Slack 👍

PaulGoldschmidt commented:
Reference also to [Issue 7 @ StanfordSpezi/spezi-web-template-application](https://github.com/StanfordSpezi/spezi-web-template-application/issues/7)

PSchmiedmayer commented:
Hope things are going well here, let me know if you need any input and how both templates are processing @PaulGoldschmidt @arkadiuszbachorski 👍

PaulGoldschmidt commented:
Over the weekend, I added the functions getUsersInformation, deleteUser and updateUserInformation to the template. Is an invitation-based registration flow, medication, appointment-system or language and timeZone in the user model part of the template? cc @PSchmiedmayer @arkadiuszbachorski 

Currently, we have in the template:
    - User types for role-based access
    - Organization relationships
    - Basic contact info
    - Core timestamps
 
Note to self: The user model currently is a simple Zod schema instead of layered schemas relying on lazy converters as no inheritance from UserRegistration is given currently.

PaulGoldschmidt committed:
more basic functions, fhir support to be replaced by basic package

PSchmiedmayer commented:
@PaulGoldschmidt Thank you for working on this! 

I think that we don't need an invitation-code based system for the template, that might be too complicated to maintain as the base template. But I can see that some of our Spezi-firebase-related packages might provide nice out-of-the-box functionality to implement such a system and or a nice "getting started" documentation article that points users in the ways they could implement such a system.

Medications and Appointments don't need to be part of the main template flow; they can be simple additive changes and developers could inspect ENGAGE-HF and others to see how this works. We have already discussed with @arkadiuszbachorski that we should have a nice npm package in the future that contains common frontend components for medication entry, lab value entry, and more that build on FHIR models & simplify the management of these in the backend and they should be easily applicable to the accompanying frontend template and this backend one if one needs them for their project (here they would probably only need to change security rules ...).

Language and time zone would be nice to have on the user model, this can be combined with the APNS token upload or similar functionality.

PaulGoldschmidt commented:
good point in keeping this as minimal as possible. I created an issue for the new package in [StanfordSpezi/spezi-firebase](https://github.com/StanfordSpezi/spezi-firebase/issues/30), maybe down the line (sometime™) it could be cool to do a show-off spinoff expanding on the template implementing every spezi-firebase package as a best-practice approach, IMO new developers could be helped greatly with just implementing the things we build once and let them strip away code from there instead of building from the docs for our packages.

I worked today on updating the user model to include lang and tz and will be doing a review of the current state tomorrow with @pauljohanneskraft and @arkadiuszbachorski 🚀

PaulGoldschmidt committed:
docs updated

PaulGoldschmidt committed:
update user model

add tz & lang

arkadiuszbachorski commented on functions/package.json:
```diff
@@ -0,0 +1,44 @@
+    "test": "echo \"No tests specified\" && exit 0",
+    "test:emulators": "echo \"No emulator tests specified\" && exit 0"
+  },
+  "engines": {
+    "node": "18"
+  }
```
@PaulGoldschmidt Would be great to bump Node version to v22. v18 reached End of Life.

PaulGoldschmidt replied:
updated in [commits/5ab54bcf80bb7531a6bea5ab6e850e4fb0c0e94d](https://github.com/StanfordSpezi/spezi-firebase-template/pull/2/commits/5ab54bcf80bb7531a6bea5ab6e850e4fb0c0e94d)

arkadiuszbachorski committed:
Use spezi-web-configurations for ESLint (#2)

# Use spezi-web-configurations for ESLint

## :recycle: Current situation & Problem
This PR introduces new Node config of spezi-web-configuration to the
template. Most of the changes have been automatically generated, some I
applied manually. There are still ESLint errors on my machine, but
mostly caused by `any` and non-existing
`@stanfordbdhg/spezi-firebase-models`. I am not sure If I am unable to
build it properly or is it not ready.

This PR comes with two commits that shouldn't land in the `main` branch:
[temp! Install local
configurations](https://github.com/StanfordSpezi/spezi-firebase-template/commit/50b086c75514d899803df7538485a8d08c15c1e6)
[temp! install tailwind
plugin](https://github.com/StanfordSpezi/spezi-firebase-template/commit/b413a272d9ed8e64d655cacf4dbc2d1d984597a3)

Both are necessary for local development. I built and installed new
version of `spezi-web-configurations`, because it's not yet available as
a release.


## :pencil: Code of Conduct & Contributing Guidelines
By creating and submitting this pull request, you agree to follow our
[Code of
Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md)
and [Contributing
Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of
Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md)
and [Contributing
Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

---------

Co-authored-by: Paul Goldschmidt <mail@paul-goldschmidt.de>

PaulGoldschmidt commented:
alright, finally making progress here: I fixed the remaining issues from the PR #2 and merged that just now. Will do some more implementation work and we should be good to merge this branch into main by EOW.

PaulGoldschmidt commented:
also, @arkadiuszbachorski: Lets sync early next week (mon/tue?) on the current state and what needs to be done for the frontend template.

PSchmiedmayer commented:
Thank you @PaulGoldschmidt & @arkadiuszbachorski 🚀

PaulGoldschmidt committed:
Add basic workflows (#5)

# Add Workflows to the Tempate Repo

## :recycle: Current situation & Problem
Like stated in #1, we need basic workflows in the repo.

## :gear: Release Notes
Currently, the workflows are carried over from MHC.

## :books: Documentation
N/A


## :white_check_mark: Testing
N/A


## :pencil: Code of Conduct & Contributing Guidelines
By creating and submitting this pull request, you agree to follow our
[Code of
Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md)
and [Contributing
Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of
Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md)
and [Contributing
Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

PaulGoldschmidt commented:
just pushed the seeding setup, we now seed data. The approach I've chosen via the serve-seeded.mjs script, not with a cloud function like in engage-hf and mhc. The seeding is still idempotent.

PaulGoldschmidt commented:
aiming to merge this initial PR by tonight PT, just basic unit tests right this moment 🚀🚀

PaulGoldschmidt committed:
Fix Seed Data (#8)

# Fix Seed Data

## :recycle: Current situation & Problem
The approach I've chosen via the serve-seeded.mjs script, not with a
cloud function like in engage-hf and mhc. The seeding is still
idempotent 🚀.

## :gear: Release Notes

![image](https://github.com/user-attachments/assets/8a50c7b2-fc4b-47de-9007-1c614b6c8521)

![image](https://github.com/user-attachments/assets/79df170d-7b76-405c-9fd0-9d55c0c2c15c)



## :books: Documentation
*Please ensure that you properly document any additions in conformance
to [Spezi Documentation
Guide](https://github.com/StanfordSpezi/.github/blob/main/DOCUMENTATIONGUIDE.md).*
*You can use this section to describe your solution, but we encourage
contributors to document your reasoning and changes using in-line
documentation.*


## :white_check_mark: Testing
Because the script is not executed in the runtime, it is not linted nor
is (at this point) any testing on this script planned.

## :pencil: Code of Conduct & Contributing Guidelines
By creating and submitting this pull request, you agree to follow our
[Code of
Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md)
and [Contributing
Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of
Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md)
and [Contributing
Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

---------

Co-authored-by: Paul Schmiedmayer <PSchmiedmayer@users.noreply.github.com>

PSchmiedmayer commented:
Nice; thank you @PaulGoldschmidt 🚀

PSchmiedmayer commented on functions/src/functions/addStepCount.ts:
```diff
@@ -0,0 +1,89 @@
+
+type AddStepCountData = z.infer<typeof addStepCountDataSchema>;
+
+export const addStepCount = onCall(
```
Might make more sense to provide an example that parses a FHIR questionnaire instead of creating an observation? Would be great if we can link this to:
a) The Swift Spezi Template app, we should add a questionnaire there that can be parsed, would then be uploaded, and then parsed here in a way that things are also shown in the iOS app.
b) Of course connected to the web frontend, ensuring that all functions that are needed there are in place.

We might need to take a look at the linking to the Spezi Template app a bit more, we might also need to make some adjustments there, e.g., HK uploads and questionnaires + user login/onboarding.

Let's make a few follow-up issue on all of this if we want to merge a first draft first; might be distributed across different repos.

PaulGoldschmidt replied:
[Issue](https://github.com/StanfordSpezi/spezi-firebase-template/issues/9) created!

PSchmiedmayer commented on functions/eslint.config.mjs:
Can we ensure that we use the spezi web config elements here as well? Might be missing this.

PaulGoldschmidt replied:
just replaced the config.

PSchmiedmayer commented on functions/models/package.json:
```diff
@@ -0,0 +1,27 @@
+  "main": "./lib/index.js",
+  "types": "./lib/index.d.ts",
+  "dependencies": {
+    "fhir": "^4.11.1",
```
Wouldn't it make more sense to build on the spezi firebase module here & across the code base?

PaulGoldschmidt replied:
agreed. This was in there from the very start before the shared lib existed, we should use this instead 👍

PaulGoldschmidt replied:
done with #10  🚀

PSchmiedmayer reviewed (approved):
Thank you for all the work here; had some high-level comments @PaulGoldschmidt 👍

PaulGoldschmidt committed:
use spezi-web-configurations

Co-Authored-By: Paul Schmiedmayer <PSchmiedmayer@users.noreply.github.com>

PaulGoldschmidt committed:
Deprecate Self-Build Models (#10)

# Deprecate Self-Build Models

## :recycle: Current situation & Problem
At the start, we implemented all data models (following mhc and
engage-hf) on its own. This is not needed anymore as @PSchmiedmayer
pointed correctly out in #1.
all its FHIR wrappers (FHIRObservation, FHIRResource, etc.) were dead
code, and SchemaConverter was redundant with
@stanfordspezi/spezi-firebase-utils

## :gear: Release Notes
Consolidated types into functions/src/types/index.ts:
- Added User class (replacing the outdated simple interface that had
wrong field names like userType instead of type)
- Added UserAuth class + userAuthConverter
- Updated userConverter to use the correct full schema matching
Firestore documents

## :books: Documentation
Updated the readme


## :white_check_mark: Testing
N/A

## :pencil: Code of Conduct & Contributing Guidelines
By creating and submitting this pull request, you agree to follow our
[Code of
Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md)
and [Contributing
Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of
Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md)
and [Contributing
Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

PaulGoldschmidt committed:
Add basic integration tests to the project (#11)

# Add basic integration tests to the project

## :recycle: Current situation & Problem
I've created a basic test suite that checks the currently existing
functions integrated into the firebase emulator.


## :gear: Release Notes
I choose to use vitest for the testing framework (instead of simon as in
MHC), because I had a way better experience with integration testing
using this framework, also Vitest has native ESM and TypeScript support,
which makes everything easier in the setup.
We do almost exclusively integration tests here, the only unit test is
`credential.test.ts`. This way, we can test the functions in a
close-to-production-type environment.


## :books: Documentation
N/A

## :white_check_mark: Testing
see release notes

## :pencil: Code of Conduct & Contributing Guidelines
By creating and submitting this pull request, you agree to follow our
[Code of
Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md)
and [Contributing
Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of
Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md)
and [Contributing
Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

PaulGoldschmidt commented:
the REUSE status does not report currently - any idea why, @PSchmiedmayer?

PaulGoldschmidt committed:
remove package.json version update step

PaulGoldschmidt commented:
addressing the @coderabbitai comments asap.

PSchmiedmayer reviewed (approved):
Thank you for working on all of this @PaulGoldschmidt; happy to see this merged as a first version and then to align with the web frontend & iOS template app 👍

PSchmiedmayer commented:
@PaulGoldschmidt fixed the mere checks; they weren't properly configured in the settings of the repo.

PaulGoldschmidt commented on firebasestorage.rules:
```diff
@@ -0,0 +1,16 @@
+      allow read, write: if request.auth != null && request.auth.uid == userId;
+    }
+    
+    // Allow clinicians and owners to read patient files
+    match /users/{userId}/{allPaths=**} {
+      allow read: if request.auth != null && 
+        firestore.get(/databases/(default)/documents/users/$(request.auth.uid)).data.type in ['clinician', 'owner'];
```
The suggested fix references a field that doesn't exist, this is an intentional feature.

PaulGoldschmidt committed:
update firestore rules

PaulGoldschmidt commented on firestore.rules:
```diff
@@ -0,0 +1,36 @@
+      // Users can read and write their own subcollections
+      match /{document=**} {
+        allow read, write: if request.auth != null && request.auth.uid == userId;
+      }
```
The update rule on /users/{userId} (line 7) now requires `request.resource.data.type == resource.data.type`

PaulGoldschmidt commented on functions/seed.mjs:
```diff
@@ -0,0 +1,51 @@
+const usersPath = join(__dirname, "emulator-data", "auth_export", "users.json");
+const users = JSON.parse(readFileSync(usersPath, "utf-8"));
+
+const app = initializeApp({ projectId: "spezi-firebase-template" });
+const auth = getAuth(app);
+const firestore = getFirestore(app);
```
feels like this is more of a nitpick, but changed this to check first if emulator environment variables are set.

PaulGoldschmidt committed:
update seeding function

PaulGoldschmidt commented on functions/src/services/user/defaultUserService.ts:
```diff
@@ -0,0 +1,137 @@
+      docs.docs.forEach((doc) => batch.delete(doc.ref));
+    }
+
+    await batch.commit();
```
good catch. Will change this.

PaulGoldschmidt commented on spezi-firebase-template-shared:
```diff
@@ -0,0 +1 @@
+Subproject commit 79a08af4c7f34a112f76d0ac564afabb719f3507
```
Not a valid concern b/c no code that was shared to the submodule was ever needed in the tests, but we can deprecate the submodule altogether. Ill do that rightaway.

PaulGoldschmidt committed:
rm submodule

PaulGoldschmidt committed:
use recursiveDelete in user document deletoin

PaulGoldschmidt commented:
@coderabbitai addressed the changes, pls review.

PaulGoldschmidt committed:
type checking issue

PaulGoldschmidt committed:
Narrowed the signature of updateUserInfo from Partial<User> to Partial<UpdatableUserInfo>

PaulGoldschmidt commented on functions/src/services/user/defaultUserService.ts:
```diff
@@ -0,0 +1,119 @@
+    });
+  }
+
+  async updateAuth(userId: string, auth: Partial<UserAuth>): Promise<void> {
+    const updateRequest: {
+      displayName?: string;
+      email?: string;
+      emailVerified?: boolean;
+      disabled?: boolean;
+      phoneNumber?: string;
+    } = {};
+    if (auth.displayName !== undefined)
+      updateRequest.displayName = auth.displayName;
+    if (auth.email !== undefined) updateRequest.email = auth.email;
+    if (auth.emailVerified !== undefined)
+      updateRequest.emailVerified = auth.emailVerified;
+    if (auth.disabled !== undefined) updateRequest.disabled = auth.disabled;
+    if (auth.phoneNumber !== undefined)
+      updateRequest.phoneNumber = auth.phoneNumber;
+
+    await getAuth().updateUser(userId, updateRequest);
+
+    if (auth.customClaims !== undefined) {
+      await getAuth().setCustomUserClaims(userId, auth.customClaims);
+    }
+  }
```
The architecture intentionally separates Auth and Firestore as distinct data stores.

PaulGoldschmidt committed:
remove broken link

arkadiuszbachorski commented on functions/eslint.config.mjs:
```diff
@@ -0,0 +1,36 @@
+  {
+    files: ["**/*.{ts,tsx}"],
+    rules: {
+      "@typescript-eslint/no-explicit-any": "warn",
```
Why is `no-explicit-any` a warning? I found it one of the crucial rules in TS-ESLint - to ensure there's no explicit `any` at all. 

It's better to replace explicit any with:
- unknown, 
- proper type, 
- run-time validation 
- ts-expect-error with a comment explaining why
- one-time ESLint rule ignore with a comment explaining why

All of these bring more context rather than `any`, which introduces possibility for reduced type safety and shares no context.

arkadiuszbachorski commented on functions/eslint.config.mjs:
```diff
@@ -0,0 +1,36 @@
+    files: ["**/*.{ts,tsx}"],
+    rules: {
+      "@typescript-eslint/no-explicit-any": "warn",
+      "@typescript-eslint/no-non-null-assertion": "warn",
```
I was wondering if we should keep `no-non-null-assertion` in general for the config, but decided to keep it purposefully.

Often, it's better to simply add a run-time check to exclude nulls rather than type annotations. 

`!` at the end of a statement is really a subtle thing to notice and sometimes can be confusing.

Proper check guarantees run-time safety rather than compiler annotations.

arkadiuszbachorski commented on README.md:
```diff
@@ -1,2 +1,322 @@
+
+### Prerequisites
+
+- Node.js 18 or later
+- Firebase CLI (`npm install -g firebase-tools`)
+- A Firebase project
```
Wouldn't it make sense to specify the default required firebase tools version? Recently I tried to enable ENGAGE with some old version and it couldn't spawn functions. 

I updated firebase tools and it started to work.

arkadiuszbachorski commented on README.md:
```diff
@@ -1,2 +1,322 @@
+
+### Prerequisites
+
+- Node.js 18 or later
```
Why v18 or later? v18 is EoL already.  

This is a starting template for applications. It feels like suggesting the highest stable version brings the most benefits: 

- people start with the latest possible version,
- decreased chance for a need to update in the upcoming 3 months, 
- this repository doesn't have to promise that it would work on anything that EoL already

PaulGoldschmidt replied:
agreed, did not edit the Readme.md yet (since last june, see #7)

arkadiuszbachorski commented on README.md:
```diff
@@ -1,2 +1,322 @@
+- Firebase CLI (`npm install -g firebase-tools`)
+- A Firebase project
+
+### Setup
```
From the ecosystem perspective, it would amazing to have instruction how to run this seeded in the least amount of steps. 

Other apps like FE dashboard and mobile apps require Firebase BE to be run before iterating. If this process is frictionless and straightforward - more people can contribute.

Example documentation from ENGAGE-HF Web:

https://github.com/StanfordBDHG/ENGAGE-HF-Web-Frontend?tab=readme-ov-file#backend

PaulGoldschmidt replied:
ping #7

arkadiuszbachorski commented on REUSE.toml:
Can it be used for automatically generating annotations? I remember in the past I tried to create an action based on `reuse` CLI tool, but had some problems and abandoned that idea. How do you use it?

PaulGoldschmidt replied:
I use this with 

```
reuse annotate \
  --copyright="Stanford University and the project authors (see CONTRIBUTORS.md)" \
  --license="MIT" \
  --year="$(date +%Y)" \
  --copyright-prefix=spdx \
  --template=spezi \
  --merge-copyrights \
  --recursive \
  --fallback-dot-license \
  .
```

could be maybe become a github action that autoannotates, although we would have to follow a similar approach with [.github/workflows/update-versions-on-release.yml](.github/workflows/update-versions-on-release.yml) b/c we cant just push to main

arkadiuszbachorski commented on functions/src/functions/addStepCount.ts:
```diff
@@ -0,0 +1,89 @@
+
+type AddStepCountData = z.infer<typeof addStepCountDataSchema>;
+
+export const addStepCount = onCall(
```
Why is it using `onCall` rather than `validatedOnCall` from the helpers?

PaulGoldschmidt replied:
agreed 👍 this was before implementing `validatedOnCall` as I remember 🎉

arkadiuszbachorski commented on functions/src/functions/addStepCount.ts:
```diff
@@ -0,0 +1,89 @@
+  type CallableRequest,
+  HttpsError,
+} from "firebase-functions/v2/https";
+import { z } from "zod/v4";
```
`zod/v4` is no longer necessary. Functions use `4.0.5` version, which already exposes v4 using `zod` path. `zod/v4` was introduced in between the versions to make migration simpler. 

Template uses 4.0.5. Might be worth updating all the packages to the latest versions to ensure that we catch some possible regressions during development and start with latest things!

PaulGoldschmidt replied:
yes, we should update all dependencies 👍

arkadiuszbachorski commented on functions/src/helpers/validatedOnCall.ts:
```diff
@@ -0,0 +1,45 @@
+} from "firebase-functions/v2/https";
+import { z, type ZodType } from "zod/v4";
+
+export const validatedOnCall = <Schema extends ZodType, Return>(
```
Why isn't this part of the Firebase utils?

PaulGoldschmidt replied:
feels like we could spend some love updating spezi-firebase in general with the learnings from here. Created a [PR](https://github.com/StanfordSpezi/spezi-firebase/pull/99) for that.

arkadiuszbachorski commented on functions/src/functions/getUsersInformation.ts:
```diff
@@ -0,0 +1,87 @@
+
+    const result: GetUsersInformationOutput = {};
+
+    for (const userId of request.data.userIds) {
```
This is iterated sequentially. By wrapping that with `Promise.all` - it would be parallelized.

arkadiuszbachorski commented on functions/src/functions/deleteUser.ts:
```diff
@@ -0,0 +1,30 @@
+    const credential = new Credential(request.auth);
+    credential.checkOwnerOrClinician();
+
+    const databaseService = new DefaultDatabaseService(getFirestore());
+    const userService = new DefaultUserService(databaseService);
```
It feels boilerplate-ish to create these services by hand. What about a factory?

arkadiuszbachorski commented on functions/src/functions/deleteUser.ts:
```diff
@@ -0,0 +1,30 @@
+import { DefaultDatabaseService } from "../services/database/databaseService.js";
+import { DefaultUserService } from "../services/user/defaultUserService.js";
+
+const deleteUserInputSchema = z.object({
+  userId: z.string(),
+});
+
+export const deleteUser = validatedOnCall(
+  deleteUserInputSchema,
+  async (request): Promise<{ success: boolean }> => {
```
There is one thing that this template misses and it's crucial for a reliable integration with FE. 

We need to figure out how to properly share types between the systems. 

Every function needs to expose:
- input - inferred from validated input schema
- output - inferred from `validatedOnCall` result or typed explicitly

Every function needs to expose it, even if it's void, to make it possible to create/generate a reliable and consistent FE client. 

There are two real ways how we can achieve it.

ENGAGE uses separate package.
Pros: package can be versioned together with deployment versions, semantic versioning, build step can be added, clear boundaries
Cons: publishing package requires architecture, there is some cost to creating and maintaining that

I was also thinking about importing a sub-repo from this package. 
Pros: straightforward to set up and use
Cons: no semantic versioning, we can pin specific commits, less standard, boundaries between systems are less clear

ENGAGE package works great, but I am happy to explore the sub-repo path to see if my assumptions are correct. It seems like sub-repo would work pretty well for our use case, simplifying the codebases. 

What do you think?

cc @pauljohanneskraft @PSchmiedmayer

PaulGoldschmidt replied:
agreed, just threw out the seperate models package because the overhead is definitely existent. Lets sync later on if we can accomplish the same thing with a subrepo 🚀

PaulGoldschmidt replied:
I will introduce the subrepo by tonight and @arkadiuszbachorski is taking a look at how it implements with the Web Dashboard.

arkadiuszbachorski commented on functions/src/services/database/collections.ts:
```diff
@@ -0,0 +1,54 @@
+      .collection("users")
+      .doc(userId)
+      .collection("questionnaireResponses")
+      .withConverter(new FHIRDatabaseConverter(fhirObservationConverter));
```
Should this use `fhirObservationConverter`? I am not sure at this point, are questionnaire responses matching fhir observations?

arkadiuszbachorski commented on functions/src/services/auth/credential.ts:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  checkAuthenticated(): void {
+    if (!this.userId) {
+      throw new HttpsError("unauthenticated", "User is not authenticated.");
+    }
```
Constructor checks this already - could be removed.

arkadiuszbachorski commented on functions/src/index.ts:
```diff
@@ -0,0 +1,16 @@
+admin.initializeApp();
+admin.firestore().settings({ ignoreUndefinedProperties: true });
+
+export { addStepCount } from "./functions/addStepCount.js";
+export { getUserData } from "./functions/getUserData.js";
+export { onUserCreated } from "./functions/onUserCreated.js";
+export { dismissMessage } from "./functions/dismissMessage.js";
+export { registerDevice } from "./functions/registerDevice.js";
+export { unregisterDevice } from "./functions/unregisterDevice.js";
```
Not all of the functions are exposed here, meaning non all of them would be available to call.

arkadiuszbachorski reviewed (commented):
I left my review. Great to see the template becoming alive! 🚀 

Some of the comments are nit, some of them can be addressed through sub-issues.

The most crucial one for me is the types sharing patterns.

arkadiuszbachorski commented:
@PaulGoldschmidt Missing things from the initial feature scoping: https://github.com/StanfordSpezi/spezi-web-template-application/issues/7

- `dismissMessage` function exists, `dismissMessages` doesn't. We could replace `dismissMessage` with one function to solve both cases. I simply need it to support dismiss notification or dismiss all notifications feature.
- admin role and `seed data` / `static data` functions - these are functions that are meant to be called by admins only. I don't know if template needs static data, but it feels like a good concept to introduce, as it seems common. From the dashboard perspective - great example of bold gate-keeping particular features of the app for one particular type of users. Clinician vs owner is more fluid. 

Other than that and types sharing - it seems everything is there! 🚀

PSchmiedmayer commented on README.md:
```diff
@@ -1,2 +1,322 @@
+  year: 'numeric',
+  month: 'long',
+  day: 'numeric'
+})
```
Missing end of code block here:

````suggestion
})
```
````

PaulGoldschmidt commented:
1. dismissMessages can get array of Message IDs to dismiss
2. implement a admin role
3. (maybe) sync the [updateStaticData.ts](https://github.com/StanfordBDHG/ENGAGE-HF-Firebase/blob/main/functions/src/functions/updateStaticData.ts)

PaulGoldschmidt commented:
Current status: There are a few PRs in related repos ([StanfordSpezi/spezi-firebase-template-shared](https://github.com/StanfordSpezi/spezi-firebase-template-shared/pull/11), [StanfordSpezi/spezi-firebase](https://github.com/StanfordSpezi/spezi-firebase/pull/99)) that should be merged before merging this PR. Happy for any comments on this PR and these efforts.

PaulGoldschmidt committed:
Update README.md

Co-authored-by: Paul Schmiedmayer <PSchmiedmayer@users.noreply.github.com>
Signed-off-by: Paul Goldschmidt <github@goldschmidt.id>

arkadiuszbachorski commented:
@PaulGoldschmidt is `emulator-data` purposefully not versioned? I was trying to run this locally. I ran `npm run prepare` and `npm run serve:seeded` in the root. 

Got this error:

```
Error: Directory "/Users/arkadiuszbachorski/Projekty/spezi-firebase-template/functions/emulator-data" does not exist.
node:internal/errors:985
  const err = new Error(message);
              ^

Error: Command failed: firebase emulators:exec --only auth,firestore,storage --import=./emulator-data --export-on-exit=./.emulator-cache "node seed.mjs"
    at genericNodeError (node:internal/errors:985:15)
    at wrappedFn (node:internal/errors:539:14)
    at checkExecSyncError (node:child_process:925:11)
    at execSync (node:child_process:997:15)
    at file:///Users/arkadiuszbachorski/Projekty/spezi-firebase-template/functions/serve-seeded.mjs:29:3
    at ModuleJob.run (node:internal/modules/esm/module_job:377:25)
    at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:691:26)
    at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:101:5) {
  status: 1,
  signal: null,
  output: [ null, null, null ],
  pid: 23864,
  stdout: null,
  stderr: null
}
```

I generated missing seed data based on schemas and the script.

Is there some way to generate seed data? I'd love firebase template to be runnable with just these two commands, with rich seeded data already.
